File: best-time-to-buy-and-sell-stock/njngwn.java
// Time Complexity: O(n), n: prices.length
// Space Complexity: O(1), because 3 int-variables are declared -> O(1)+O(1)+O(1)
class Solution {
    public int maxProfit(int[] prices) {
        int buyPrice = prices[0];
        int sellPrice = 0;
        int profit = 0;  // store maxProfit

        for (int i = 1; i < prices.length; ++i) {
            buyPrice = Math.min(buyPrice, prices[i]);
            profit = Math.max(profit, prices[i] - buyPrice);
        }

        return profit;
    }
}

File: best-time-to-buy-and-sell-stock/njngwn.py
class Solution:
    # Time Complexity: O(n), n: len(prices)
    # Space Complexity: O(1)
    def maxProfit(self, prices: List[int]) -> int:
        min_price = prices[0]
        profit = 0

        for i in range(1, len(prices)):
            min_price = min(min_price, prices[i])
            profit = max(profit, prices[i] - min_price)

        return profit
